src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -222,4 +239,226 @@ begin
+  obtain ⟨ry, sy, hy⟩ := is_localization.mk'_surjective (non_zero_divisors R) y,
+  rw [← hxy, ← hx, ← hy, valuation_of_mk', valuation_of_mk', valuation_of_mk'],
+  have h_frac_xy : is_localization.mk' K rxy sxy =
+    is_localization.mk' K (rx*(sy : R) + ry*(sx : R)) (sx*sy),
+  { rw [is_localization.mk'_add, hx, hy, hxy], },
+  have h_frac_x : is_localization.mk' K rx sx = is_localization.mk' K (rx*(sy : R)) (sx*sy),
+  { rw [is_localization.mk'_eq_iff_eq, submonoid.coe_mul, mul_assoc, mul_comm (sy : R) _], },
+  have h_frac_y : is_localization.mk' K ry sy = is_localization.mk' K (ry*(sx : R)) (sx*sy),
+  { rw [is_localization.mk'_eq_iff_eq, submonoid.coe_mul, mul_assoc], },
+  have h_denom : 0 < v.int_valuation_def ↑(sx * sy),
+  { rw [int_valuation_def, if_neg _],
+    { exact with_zero.zero_lt_coe _ },
+    { exact non_zero_divisors.ne_zero
+        (submonoid.mul_mem (non_zero_divisors R) sx.property sy.property), }},
+  rw [int_valuation_div_eq_div K v h_frac_x, int_valuation_div_eq_div K v h_frac_y,
+    int_valuation_div_eq_div K v h_frac_xy, le_max_iff, div_le_div_right₀ (ne_of_gt h_denom),
+    div_le_div_right₀ (ne_of_gt h_denom), ← le_max_iff],
+  exact v.int_valuation.map_add_le_max' _ _,
+end
+
+/-- The `v`-adic valuation on `K`. -/
+def valuation  : valuation K (with_zero (multiplicative ℤ)) :=
+{ to_fun    := v.valuation_def,
+  map_zero' := valuation.map_zero' v,
+  map_one'  := valuation.map_one' v,
+  map_mul'  := valuation.map_mul' v,
+  map_add_le_max'  := valuation.map_add_le_max' v }
+
+variable (K)
+/-- There exists `π ∈ K` with `v`-adic valuation `multiplicative.of_add (-1)`. -/
+lemma valuation_exists_uniformizer :
+  ∃ (π : K), v.valuation_def π = multiplicative.of_add (-1 : ℤ) :=
+begin
+  obtain ⟨r, hr⟩ := v.int_valuation_exists_uniformizer,
+  use algebra_map R K r,
+  rw valuation_of_algebra_map v,
+  exact hr,
+end
+
+/-- Uniformizers are nonzero. -/
+lemma valuation_uniformizer_ne_zero :
+  (classical.some (v.valuation_exists_uniformizer K)) ≠ 0 :=
+begin
+  have hu := (classical.some_spec (v.valuation_exists_uniformizer K)),
+  have h : v.valuation_def (classical.some _) = valuation v (classical.some _) := rfl,
+  rw h at hu,
+  exact (valuation.ne_zero_iff _).mp (ne_of_eq_of_ne hu with_zero.coe_ne_zero),
+end
+
+/-! ### Completions with respect to adic valuations
+Given a Dedekind domain `R` with field of fractions `K` and a maximal ideal `v` of `R`, we define
+the completion of `K` with respect to its `v`-adic valuation, denoted `v.adic_completion`, and its
+ring of integers, denoted `v.adic_completion_integers`. -/
+
+variable {K}
+
+/-- `K` as a valued field with the `v`-adic valuation. -/
+def v_valued_K : valued K (with_zero (multiplicative ℤ)) := ⟨v.valuation⟩
+
+lemma v_valued_K_def {x : K} : @valued.v K _ _ _ (v_valued_K v) (x) = v.valuation_def x := rfl
+
+/-- The topological space structure on `K` corresponding to the `v`-adic valuation. -/
+def v_valued_K_topological_space : topological_space K :=
```

Comment:
Can you add a docstring saying why this cannot be an instance? Thanks!

Reply:
Would "This cannot be made an instance since it depends on the choice of `v`." work? (I think the issue is that we should not have multiple topological space structures on `K`).

Reply:
Yes, something like that is perfect. And I totally agree it cannot be an instance.